File: JAVA/StringToInteger.java
class Solution {
	public boolean isDigit(char ch) {
		return ch >= '0' && ch <= '9';
	}

	public int atoi(String str) {
		// Start typing your Java solution below
		// DO NOT write main() function
		if (null == str || str.length() == 0)
			return 0;
		int i = 0;
		while (i < str.length() && str.charAt(i) == ' ')
			i++;
		if (i == str.length())
			return 0;
		StringBuilder digits = new StringBuilder();
		char first = str.charAt(i);
		if (first == '+' || first == '-')
			i++;
		else if (!isDigit(first))
			return 0;
		while (i < str.length() && isDigit(str.charAt(i))) {
			digits.append(str.charAt(i));
			i++;
		}
		if (digits.length() == 0)
			return 0;

		return  StringToInt(digits.toString(),first);
	}

	private int StringToInt(String s, char sign) {
		if(s.length() > 10)
			return sign == '-' ? Integer.MIN_VALUE : Integer.MAX_VALUE;
		int sum = 0;
		for (int i = 0; i < s.length(); i++) {
			sum *= 10;
			sum += s.charAt(i) - '0';
			if(sign != '-' && sum < 0)
				return Integer.MAX_VALUE;
			else if(sum < 0)
				return Integer.MIN_VALUE;
		}
		return sign == '-' ? -sum : sum;
	}
}

public class StringToInteger {
	public static void main(String[] args)
	{
		Solution sol = new Solution();
		System.out.println(""+sol.atoi("    10522545459"));
	}
}

File: C++/StringToInteger.cpp
#include<iostream>
#include<stdlib.h>
#include<vector>
#include<algorithm>
#include<math.h>
#include<map>
using namespace std;
const int intmax = 2147483647;
const int intmin = -2147483648;
class Solution {
public:

	bool isNum(char ch)
	{
		if(ch >= '0' && ch <= '9')
			return true;
		return false;
	}
	int atoi(const char *str) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function
		if(!str)
			return 0;
		int len = strlen(str);
		if(len == 0)
			return 0;
		int i = 0;
		while(str[i] == ' ')
			i++;
		if(!str[i])
			return 0;
		bool sign = true;
		if(str[i] == '-' || str[i] == '+')
		{
			sign = str[i] == '+' ? true : false;
			i++;
		}
		else if(!isNum(str[i]))
			return 0;
		int sum = 0;
		int count = 0;
		while(isNum(str[i]))
		{
			sum *= 10;
			sum += str[i] - '0';
			i++;
			count++;
			if(count > 10 || sum < 0)
				return sign ? intmax : intmin;
		}
		return sign ? sum:-sum;
	}
};
int main()
{
	Solution sol;
	cout<<sol.atoi("10557266014");
	getchar();
}